### PR TITLE
Titan: Remove superversion dependency

### PR DESCRIPTION
**Pull request by Connor1996**
remove titan version mechanism to avoid superversion dependency

### CODE REVIEW

File: utilities/titandb/db_impl_files.cc [hunk 1]
```diff
@@ -3,17 +3,20 @@
...
 namespace titandb {
 
-void TitanDBImpl::PurgeObsoleteFiles() {
+std::pair<bool, int64_t> TitanDBImpl::PurgeObsoleteFiles(bool aborted) {
   Status s;
   ObsoleteFiles obsolete_files;
-  vset_->GetObsoleteFiles(&obsolete_files);
+  {
+    MutexLock l(&mutex_);
+    auto oldest_sequence = GetOldestSnapshotSequence();
```
Comment (yiwu-arbug) [inline utilities/titandb/db_impl_files.cc]: It seems to me `GetOldestSnapshotSequence()` can be called without holding `mutex_`.

Comment (Connor1996) [reply utilities/titandb/db_impl_files.cc]:
```
  // The lock sequence must be Titan.mutex_.Lock() -> Base DB mutex_.Lock()
  // while the unlock sequence must be Base DB mutex.Unlock() ->
  // Titan.mutex_.Unlock() Only if we all obey these sequence, we can prevent
  // potential dead lock.
  port::Mutex mutex_;
```
We will hold base_db_lock in `GetOldestSnapshotSequence`, so get Titan's lock here

Comment (yiwu-arbug) [reply utilities/titandb/db_impl_files.cc]: I think the comment only applies when we want to lock both of TitanDB mutex and base DB's mutex simultaneously? In this case we release one lock before locking the other, and I think it is fine.

File: utilities/titandb/db_impl.cc [hunk 2]
```diff
@@ -233,6 +244,13 @@ Status TitanDBImpl::CloseImpl() {
     }
   }
 
+  if (thread_purge_obsolete_ != nullptr) {
+    thread_purge_obsolete_->cancel();
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply utilities/titandb/db_impl.cc]: https://github.com/pingcap/rocksdb/blob/08f3371694a9f833b3fb354ff599685214eab73b/utilities/titandb/version_set.cc#L110
https://github.com/pingcap/rocksdb/blob/08f3371694a9f833b3fb354ff599685214eab73b/utilities/titandb/version_set.cc#L127
already done here

**Comment by Connor1996** [comment]
@yiwu-arbug Yes, I will add later

File: utilities/titandb/db_impl_files.cc (hunk 1 above)
Comment (tabokie) [reply utilities/titandb/db_impl_files.cc]: Agree, removing that mutex won't lead to a possible deadlock.

File: utilities/titandb/db_impl.cc [hunk 3]
```diff
@@ -295,6 +295,11 @@ Status TitanDBImpl::DropColumnFamilies(
 
   MutexLock l(&mutex_);
 
+  // TODO:
+  // As rocksdb described, `DropColumnFamilies()` only records the drop the column family specified by ColumnFamilyHandle.
```
Comment (tabokie) [inline utilities/titandb/db_impl.cc]: `only records the drop the column family`, meaning is unclear, probablly a typo.

**Comment by tabokie** [comment]
Looks good to me!

File: utilities/titandb/db_impl.cc [hunk 4]
```diff
@@ -118,6 +118,17 @@ TitanDBImpl::TitanDBImpl(const TitanDBOptions& options,
 
 TitanDBImpl::~TitanDBImpl() { Close(); }
 
+// how often to schedule delete obs files periods
+static constexpr uint32_t kDeleteObsoleteFilesPeriodSecs = 10; // 10s
+
+void TitanDBImpl::StartBackgroundTasks() {
+  if (!thread_purge_obsolete_) {
+    thread_purge_obsolete_.reset(new rocksdb::RepeatableThread(
+        [this]() { TitanDBImpl::PurgeObsoleteFiles(); }, "purge", env_,
```
[the remark this answers is not in the document]
Comment (Connor1996) [reply utilities/titandb/db_impl.cc]: RepeatableThread will add thread name with prefix `rocksdb:`, the length limit is 16, so I change this into `rocksdb:titanbg`

**Comment by Connor1996** [comment]
PTAL again~ @yiwu-arbug

**Comment by yiwu-arbug** [comment]
@Connor1996 hey, you stay up so late! sorry for a bit delay. I'll finish review today.

File: utilities/titandb/blob_gc_job_test.cc [hunk 5]
```diff
@@ -68,7 +73,13 @@ class BlobGCJobTest : public testing::Test {
...
   }
 
-  void DestoyDB() { db_->Close(); }
+  void Flush() {
+    FlushOptions fopts;
+    fopts.wait = true;
+    ASSERT_OK(db_->Flush(fopts));
+  }
+
+  void DestroyDB() { db_->Close(); }
```
Comment (yiwu-arbug) [inline utilities/titandb/blob_gc_job_test.cc]:
```
auto s __attribute__((__unused__)) = db_->Close();
assert(s.ok());
delete db_;
```

File: utilities/titandb/blob_gc_job_test.cc [hunk 6]
```diff
@@ -214,14 +220,67 @@ class BlobGCJobTest : public testing::Test {
...
-    DestoyDB();
+    DestroyDB();
   }
 };
 
 TEST_F(BlobGCJobTest, DiscardEntry) { TestDiscardEntry(); }
 
 TEST_F(BlobGCJobTest, RunGC) { TestRunGC(); }
 
+TEST_F(BlobGCJobTest, PurgeBlobs) {
```
Comment (yiwu-arbug) [inline utilities/titandb/blob_gc_job_test.cc]: Add one or two sentences of inline comment to describe the flow of the test would be better :)

**Comment by yiwu-arbug** [comment]
There's some compile errors when running `make titandb_check`. Also mind also update `TitanDBTest::GetBlobStorage`? It don't need to get snapshot and ref version after your change. Thanks. https://github.com/pingcap/rocksdb/blob/tikv-3.0/utilities/titandb/titan_db_test.cc#L111

**Comment by Connor1996** [comment]
Sorry for the delay, busy working on other things these days. seems the conflict is neglected when merging branch, I will fix it.

**Comment by Connor1996** [comment]
Done. And I will make some polish in later PR.

**Comment by yiwu-arbug** [comment]
@Connor1996 Thanks!

**Pull request merged**